Fix SuperIlc to work on Linux / OSX in composite mode

trylek opened the pull request:
1) I have generalized the concept of framework exclusions so that
the CPAOT runner can consume it - without it we were picking up
"CommandLine.dll" in the Core_Root folder as a Crossgen2 compilation
input and R2RDump was subsequently failing to dump the file as
CommandLine.dll is one of the stupid assemblies where the assembly
name differs from the PE file name (the assembly name is
'commandline' causing a mismatch on case-sensitive file systems).

2) On Linux / OSX, we need to copy around a bunch of native libraries
to the output, e.g. libSystem.Globalization.Native.so.

With this change, SuperIlc seems to be successfully running a small
set of hand-picked tests. I plan to follow up by preparing the PR
for introducing Crossgen2 composite pipeline jobs to get a more
precise reading on the pass rate.

Thanks

Tomas

trylek committed:
Fix SuperIlc to work on Linux / OSX in composite mode

1) I have generalized the concept of framework exclusions so that
the CPAOT runner can consume it - without it we were picking up
"CommandLine.dll" in the Core_Root folder as a Crossgen2 compilation
input and R2RDump was subsequently failing to dump the file as
CommandLine.dll is one of the stupid assemblies where the assembly
name differs from the PE file name (the assembly name is
'commandline' causing a mismatch on case-sensitive file systems).

2) On Linux / OSX, we need to copy around a bunch of native libraries
to the output, e.g. libSystem.Globalization.Native.so.

With this change, SuperIlc seems to successfully running a small
set of hand-picked tests. I plan to follow up by preparing the PR
for introducing Crossgen2 composite pipeline jobs to get a more
precise reading on the pass rate.

Thanks

Tomas

janvorli commented on src/coreclr/src/tools/ReadyToRun.SuperIlc/CpaotRunner.cs:
```diff
@@ -116,11 +119,15 @@ private List<string> ResolveReferences()
             char referenceOption = (_options.Composite ? 'u' : 'r');
             List<string> references = new List<string>();
-            foreach (var referenceFolder in _referenceFolders)
+            foreach (string referenceFolder in _referenceFolders)
```
A nit - It looks like it would be nice to unify this with the same code starting at line 97.

trylek replied:
Good point, thanks. Will do!

trylek committed:
Unify code per JanV's PR feedback; bump up composite build timeout

trylek committed:
Further simplify reference generator to avoid emitting duplicates

trylek commented:
@janvorli, I have refactored CpaotRunner based on your suggestion in two steps - my first change got rid of the code duplication but turned out to be problematic for composite shared library build as it ended up producing duplicate input files crashing Crossgen2. While fixing this I have realized I can simplify the emitter even further. I have also bumped up ILC timeout in composite mode to 30 minutes because, when running under debug-built corerun, it's slow as hell. Can you please take another look when you have a chance?

Thanks

Tomas

trylek committed:
Fix two more bugs in my change I found in local testing

1) Stupid me, I reversed the polarity of two FrameworkExclusion.Excluded
checks.

2) Apparently my earlier change for composite support refactoring
ProcessRunner to accept multiple input files broke "normal R2R"
framework compilation (orthogonal to composite shared framework);
this additional delta fixes that too.

Thanks

Tomas
